groups/reconcile.go

```diff
@@ -0,0 +1,397 @@
+package main
+
+import (
+	"crypto/tls"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/googleapi"
+	"google.golang.org/api/groupssettings/v1"
+	"gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	// the email id for the bot/service account
+	BotID string `yaml:"bot-id"`
+
+	// the file with the authentication information
+	TokenFile string `yaml:"token-file,omitempty"`
+
+	// the file with the groups/members information
+	GroupsFile string `yaml:"groups-file,omitempty"`
+
+	// If true, don't make any mutating API calls
+	DryRun bool
+}
+
+type GroupsConfig struct {
+	// List of google groups
+	Groups []GoogleGroup `yaml:"groups,omitempty"`
+}
+
+type GoogleGroup struct {
+	EmailId     string `yaml:"email-id"`
+	Description string `yaml:"description"`
+
+	// List of members in the google group
+	Members []string `yaml:"members,omitempty"`
+}
+
+func Usage() {
+	fmt.Fprintf(os.Stderr, `
+Usage: %s [-config <config-yaml-file>] [-dry-run]
+Command line flags override config values.
+`, os.Args[0])
+	flag.PrintDefaults()
+}
+
+var config Config
+var groupsConfig GroupsConfig
+
+func main() {
+	configFilePath := flag.String("config", "config.yaml", "the config file in yaml format")
+	dryRun := flag.Bool("dry-run", false, "do not push anything to google groups")
+
+	flag.Usage = Usage
+	flag.Parse()
+
+	err := readConfig(configFilePath, dryRun)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	err = readGroupsConfig(config.GroupsFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	jsonCredentials, err := ioutil.ReadFile(config.TokenFile)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	credential, err := google.JWTConfigFromJSON(jsonCredentials, admin.AdminDirectoryUserReadonlyScope,
+		admin.AdminDirectoryGroupScope,
+		admin.AdminDirectoryGroupMemberScope,
+		groupssettings.AppsGroupsSettingsScope)
+	if err != nil {
+		log.Fatalf("Unable to parse client secret file to config: %v\n. "+
+			"Please run 'git-crypt unlock'", err)
+	}
+	credential.Subject = config.BotID
+
+	client := credential.Client(context.Background())
+	srv, err := admin.New(client)
+	if err != nil {
+		log.Fatalf("Unable to retrieve directory Client %v", err)
+	}
+
+	srv2, err := groupssettings.New(client)
+	if err != nil {
+		log.Fatalf("Unable to retrieve groupssettings Service %v", err)
+	}
+
+	log.Println(" =================== Current Status ======================")
+	err = printGroupMembersAndSettings(srv, srv2)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println(" ======================= Updates =========================")
+	for _, g := range groupsConfig.Groups {
+		if !strings.HasPrefix(g.EmailId, "k8s-infra-") {
+			log.Fatalf("We can reconcile only groups that start with 'k8s-infra-' prefix")
```

Comment:
Doesn't have to be now, but would like to see k8s-infra- prefix extracted

Reply:
Added to follow up issue https://github.com/kubernetes/k8s.io/issues/255